### ISSUE TITLE
Find a way to pass the same CMAKE_TOOLCHAIN_FILE to every cmake-based subproject

### ISSUE THREAD

**Issue by traversaro** [robotology/ycm#129]
For cross-compilation of a complete workspace, it could be convenient to be able to pass in some way the same `CMAKE_TOOLCHAIN_FILE` to every cmake-based subproject of a YCM-based superbuild. 

An alternative solution that could be implemented in CMake would be supporting specifying `CMAKE_TOOLCHAIN_FILE` as an environment variable. I remember that this solution was suggested somewhere in vcpkg docs, but I cannot find any reference for this. However, I think this was never proposed/discussed in the CMake mailing list/issue tracker, and it may be undesirable as it may increase even more complexity at the CMake level. 

Related conan issue: https://github.com/conan-io/conan/issues/1447 .

**Comment by traversaro** [comment]
cc @claudiofantacci

**Comment by traversaro** [comment]
In ament, this is solved by the generic `cmake-args` option: https://github.com/ros2-for-arm/ros2/wiki/ROS2-on-arm-architecture#trigger-a-build .

**Comment by drdanz** [comment]
I think that the easiest way to do that is to pass it here:
https://github.com/robotology/ycm/blob/master/modules/YCMEPHelper.cmake#L790

**Comment by traversaro** [comment]
Indeed, the following lines : 
~~~
  if(CMAKE_TOOLCHAIN_FILE)
    list(APPEND ${_name}_YCM_CMAKE_ARGS "-DCMAKE_TOOLCHAIN_FILE:PATH=${CMAKE_TOOLCHAIN_FILE}")
  endif()
~~~
would be helpful for crosscompiling superbuilds. Would you be happy with this change @drdanz ?

**Comment by traversaro** [comment]
You meant for the iOS `robotology-superbuild` :D

**Comment by drdanz** [comment]
If it works, it looks good to me...

**Comment by traversaro** [comment]
> If it works, it looks good to me...

Ok, I will try it then. : D

**Comment by traversaro** [comment]
Related PR : https://github.com/ros2/rviz/pull/374 .

**Comment by traversaro** [comment]
Example of this in the wild: 
* https://code.videolan.org/videolan/medialibrary/commit/5c16352cf4cc945f9d034b45be5e427b6064c4e2 
* https://github.com/xbmc/xbmc/blob/eb59376472d55a4006a98635a2f98990e4fb4cb1/cmake/addons/CMakeLists.txt#L118

**Comment by cherullo** [comment]
Hi,

Just dropping by to say that @drdanz's fix is required and works when building gazebo for Windows using vcpkg and @traversaro's [gazebo-superbuild](https://github.com/traversaro/gazebo-superbuild).

Is there any way I can help?

**Comment by traversaro** [comment]
Feel  free to open a PR with that fix @cherullo (btw, why are toolchains with `gazebo-superbuild`? vcpkg?).

**Comment by traversaro** [comment]
If you are compiling Gazebo on Windows, pay attention that the fix for the issue described in https://github.com/ms-iot/ROSOnWindows/issues/47 still needs to be merged upstream.

**Comment by cherullo** [comment]
I'm starting the compilation for Windows from scratch to better document the process.
I'll create the pull request as soon as I get to this issue again (which won't be long now).
Thanks for the heads-up on the IP issue!

**Comment by drdanz** [comment]
Fixed by #208 
Thanks @cherullo
